### PR TITLE
tailor: fix #3093, #3103

### LINKED ISSUE #3093: tailor/autoclothing(?) -- ordering too many clothes due to strange circumstances
Save: https://pixeldrain.com/l/kLSdTo1i (ignore mods2.zip)

Currently I have tailor enabled, and autoclothing is set to 1 silk cloak and 1 silk gloves.

This fort is for some reason producing excessive surpluses of trousers and dresses and a somewhat lesser but still alarming surplus of socks.  I have two particular residents which I know must be responsible for a large portion of the issue: 

- One (Tobul Kacothiden, militia captain vampire) is a vampire I have kept in isolation for some time with her own facilities to make new clothing.  Tailor seems to still be allocating clothing to her, understandably -- I don't really expect it to check for someone being quarantined.  It also, according to the logs, seems to be generating orders for new clothing for covering her *every day* she has worn footwear, bodywear, or legwear.

- Another (Amkash Rodohthremlu, Elephant Man Scholar risen hunter) is an elephant man implicated by the large surplus of elephant man sized clothing.  He seems to be bugged in a lot of ways since when other people would drop an item, he continues to haul it.  Before I raised him as undead he would also refuse to drop mugs, and would refuse to eat and would instead starve to the point of passing out and get food delivered to him (Putnam suspects this is due to the GRAZER tag being kept for animal men).  I suspect this could be causing larger than normal delays in acquiring new clothing and therefore generating too many orders.

Autoclothing is able to detect some of the surplus clothing, but the accounting is all wrong.  It says I have 30 legwear for dwarves, but from my count of my stockpile for trousers I have somewhere in the ballpark of 300 trousers, unworn, for various species.  It sounds like clothing is being continuously allocated whenever it isn't picked up, and that is leading to duplicate orders since the allocation of stockpiled clothing isn't cleared.

### LINKED ISSUE #3103: Using makeown and tailor together gives weird results
If you use makeown on an animal or other creature that does not feel a need for clothes tailor will still happily produce clothes for them.

I believe this could be solved by tailor checking for the `CAN_LEARN` caste flag in addition to the current checks here:
https://github.com/DFHack/dfhack/blob/253832400883474f710291ef0b35bc4e63f8b171/plugins/tailor.cpp#L273-L277

### PR DESCRIPTION
**Pull request by ab9rf**
I significantly restructured the assignment algorithm to eliminate the overproduction issues in #3093

#3103 is addressed by excluding units that don't have the caste-level EQUIPS flag

closes #3093
closes #3103

### CODE REVIEW

File: plugins/tailor.cpp
Comment (myk002) [inline plugins/tailor.cpp:1]: gcc complains about unused vars:
```
/home/runner/work/dfhack/dfhack/plugins/tailor.cpp: In member function ‘void Tailor::scan_replacements()’:
/home/runner/work/dfhack/dfhack/plugins/tailor.cpp:291:22: error: unused variable ‘allocated’ [-Werror=unused-variable]
  291 |                 bool allocated = false;
      |                      ^~~~~~~~~
/home/runner/work/dfhack/dfhack/plugins/tailor.cpp: In member function ‘void Tailor::scan_existing_orders()’:
/home/runner/work/dfhack/dfhack/plugins/tailor.cpp:407:18: error: unused variable ‘sub’ [-Werror=unused-variable]
  407 |             auto sub = o->item_subtype;
      |                  ^~~
```

File: docs/changelog.txt [hunk 1]
```diff
@@ -44,7 +44,7 @@ changelog.txt uses a syntax similar to RST, with a few special sequences:
 - `buildingplan`: you can no longer designate constructions on tiles with magma or deep water
 - `buildingplan`: fixed material filter getting lost for planning buildings on save/reload
 - `buildingplan`: respect building size limits (e.g. roads and bridges cannot be more than 31 tiles in any dimension)
-- `tailor`: now properly discriminates between dyed and undyed cloth, no longer defaults to using adamantine, and properly tracks material requirements for already queued orders
+- `tailor`: now properly discriminates between dyed and undyed cloth, no longer defaults to using adamantine, properly tracks material requirements for already queued orders, skips units who can't wear clothes, and hopefully won't over-order items anymore
```
Comment (myk002) [inline docs/changelog.txt:47]: it's fine to have multiple `tailor` lines -- they'll be formatted in a sublist by the changelog parser

File: plugins/tailor.cpp
Comment (ab9rf) [reply plugins/tailor.cpp:1]: oops

File: plugins/tailor.cpp [hunk 2]
```diff
@@ -169,50 +171,34 @@ class Tailor {
...
+                continue;
+            df::item_type t = i->getType();
+            int size = world->raws.creatures.all[i->getMakerRace()]->adultsize;
 
-                available[std::make_pair(t, size)] += 1;
-            }
+            available[std::make_pair(t, size)] += 1;
+        }
 
+        for (auto& i : available)
```
Comment (myk002) [inline plugins/tailor.cpp:195]: if this loop is just for debugging, you can enclose it in an if block that tests whether the output will go anywhere:
```cpp
if (DBG_NAME(cycle).isEnabled(DebugCategory::LDEBUG))
```

**Review by myk002 — APPROVED** [review]

**Pull request merged**